### DISCUSSION TITLE
Timestamps with time zone?

### DISCUSSION THREAD

**Discussion by mdavis-xyz** [pydantic/pydantic#4700] [Pydantic V2]
One of the biggest pain points I have with v1 is parsing and serialising datetimes.

I'm dumping and loading from json objects which have various timestamp formats. Not just ISO8601. 

Most of my datetimes need to be parsed into a specific time zone (UTC+10). I generally don't want them timezone unaware.

One of the APIs I interface with actually has two datetime fields with different formats and time zones.

Currently I've tried a range of solutions, such as subclassing `dt.datetime` to add custom validators, which takes more than 100 lines of code (UTC, UTC+10, `dt.Date` and `dt.Time`) and is a continual source of frustration.

Note that sometimes timestamps end in an integer number of seconds, sometimes they have fractions of a second, and the normal `strptime` `%S` and `%f` struggles with that kind of variation.

Will Pydantic v2 have a way to solve this?

e.g.

```
class MyModel(BaseModel):
    t1: pydantic.DateTime[fmt="%Y-%m-%dT%H:%M:%S[.%f]", tzinfo=my_timezone]
```

**Comment by samuelcolvin** [comment]
Have a look at [speedate](https://github.com/pydantic/speedate) which will do the datetime parsing for V2.

But sounds like you might need your own validator.

**Reply by mdavis-xyz** [reply]
Hmm, yes I do sometimes have datetimes with slashes between year/month/day. And sometimes the timezone is not explicit (`+10:00`) but I still want to set the time zone.

Ok, I guess I'll have to keep my existing solution.

**Reply by spawn-guy** [reply]
i have an opposite question: how can i make speedate to create DateTimes TZ-Naiive?

in some places i have `timestamp=int` in `ms` (in cassandra) and `speedate` adds `tz.UTC` to the results. but i also have tz-naiive datetimes in postgres where the speedate is not triggered.

i.e. which _current_ solution is better?

parsing on my own to tz-naiive
```
class SomeBaseResponse(BaseModel):
    date_time: datetime = Field(..., alias="timestamp")

    @field_validator("date_time", mode="before")
    @classmethod
    def convert_timestamp(cls, value):
        if isinstance(value, int):
            try:
                return datetime.utcfromtimestamp(value / 1000)
            except ValueError:
                return None
        return value
```
or

parsing int with `speedate` to datetime-tz-aware and then stripping tz?
```
class SomeBaseResponse(BaseModel):
    date_time: datetime = Field(..., alias="timestamp")

    @field_validator("date_time", mode="after")
    @classmethod
    def strip_timezone(cls, value):
        if isinstance(value, datetime):
            return value.replace(tzinfo=None)
        return value
```
and this 2nd method doesn't work with Annotated
```
class SomeBaseResponse(BaseModel):
    date_time: Annotated[datetime, NaiveDatetime] = Field(..., alias="timestamp")
```
mode="before" - receives an int, which doesn't validate
mode="after" - receives a tz-aware datetime and doesn't validate
